src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -180,27 +179,32 @@ public TokenData GetNextTokenData(TokenData tokenData)
         internal SyntaxToken FirstTokenOfBaseTokenLine(SyntaxToken token)
         {
             var currentTokenData = this.GetTokenData(token);
-            while (!this.IsFirstTokenOnLine(token))
+            return FirstTokenOfBaseTokenLine(ref currentTokenData);
+        }
+
+        // using ref to avoid copying the entire struct
+        private SyntaxToken FirstTokenOfBaseTokenLine(ref TokenData tokenData)
+        {
+            while (!this.IsFirstTokenOnLine(in tokenData))
             {
-                var previousTokenData = this.GetPreviousTokenData(currentTokenData);
-                token = previousTokenData.Token;
-                currentTokenData = previousTokenData;
+                tokenData = this.GetPreviousTokenData(tokenData);
             }
 
-            return token;
+            return tokenData.Token;
         }
 
         public bool TwoTokensOriginallyOnSameLine(SyntaxToken token1, SyntaxToken token2)
         {
-            return TwoTokensOnSameLineWorker(token1, token2, _getOriginalTriviaData);
+            return TwoTokensOnSameLineWorker(token1, token2, new OriginalTriviaDataGetter(this));
         }
 
         public bool TwoTokensOnSameLine(SyntaxToken token1, SyntaxToken token2)
         {
-            return TwoTokensOnSameLineWorker(token1, token2, _getTriviaData);
+            return TwoTokensOnSameLineWorker(token1, token2, new TriviaDataGetter(this));
         }
 
-        private bool TwoTokensOnSameLineWorker(SyntaxToken token1, SyntaxToken token2, Func<TokenData, TokenData, TriviaData> triviaDataGetter)
+        private bool TwoTokensOnSameLineWorker<TDataGetter>(SyntaxToken token1, SyntaxToken token2, TDataGetter dataGetter)
+            where TDataGetter : struct, ITriviaDataGetter
```

Comment:
this is a really nifty pattern to help out with this.  tagging @YairHalberstadt for his passion around `shapes`.